src/api/app/controllers/trigger_controller.rb

```diff
@@ -1,16 +1,71 @@
 class TriggerController < ApplicationController
+  validate_action rebuild: { method: :post, response: :status }
+  validate_action release: { method: :post, response: :status }
   validate_action runservice: { method: :post, response: :status }
 
+  before_action :require_valid_token
+
   #
   # This controller is checking permission always only on the base of tokens
   #
   skip_before_action :extract_user
   skip_before_action :require_login
 
-  # github.com sends a hash payload
-  skip_before_action :validate_params, only: [:runservice]
+  def rebuild
+    Backend::Api::Sources::Package.rebuild(@pkg.project.name, @pkg.name)
+    render_ok
+  end
+
+  # rubocop:disable Lint/NonLocalExitFromIterator
+  def release
+    if params[:project].present?
+      render_error message: 'Token must define the release package',
+                   status: 403,
+                   errorcode: 'no_permission'
+      return
+    end
+
+    matched_repo = false
+    User.current = @token.user
+    @pkg.project.repositories.each do |repo|
+      repo.release_targets.each do |releasetarget|
+        next unless releasetarget.trigger == 'manual'
+        unless @token.user.can_modify?(releasetarget.repository.project)
+          render_error message: "no permission for target #{releasetarget.repository.project.name}",
+                       status: 403,
+                       errorcode: 'no_permission'
+          return
+        end
+        target_package_name = @pkg.release_target_name
```

Comment:
This is the same every time you iterate, you should put it outside the iteration (small performance improvement)

Reply:
^^

Reply:
This is the same every time you iterate, you should put it outside the iteration (small performance improvement)

Reply:
this haven't been addressed

Reply:
AFAIR in your original change request, you handled it as nitpick. So, because Adrian didn't want to change it, I will keep it too (probably because the number of release_targets where trigger = manual won't end up in thousand lines iteration)